libazureinit/src/distro.rs

```diff
@@ -39,58 +26,44 @@ impl Distribution for Distributions {
                     .arg(home_path.clone())
                     .arg("-m")
                     .status()?;
-                if !status.success() {
-                    return Err(Error::SubprocessFailed {
-                        command: "useradd".to_string(),
-                        status,
-                    });
-                }
-
-                if password.is_empty() {
-                    let status = Command::new("passwd")
-                        .arg("-d")
-                        .arg(username)
-                        .status()?;
-                    if !status.success() {
-                        return Err(Error::SubprocessFailed {
-                            command: "passwd".to_string(),
-                            status,
-                        });
-                    }
-                } else {
-                    // creating user with a non-empty password is not allowed.
-                    return Err(Error::NonEmptyPassword);
-                }
-
-                Ok(0)
-            }
-        }
+    if !status.success() {
+        return Err(Error::SubprocessFailed {
+            command: "useradd".to_string(),
+            status,
+        });
     }
-    fn set_hostname(&self, hostname: &str) -> Result<i32, Error> {
-        match self {
-            Distributions::Debian | Distributions::Ubuntu => {
-                let status = Command::new("hostnamectl")
-                    .arg("set-hostname")
-                    .arg(hostname)
-                    .status()?;
-                if status.success() {
-                    Ok(status.code().unwrap_or(1))
-                } else {
-                    Err(Error::SubprocessFailed {
-                        command: "chpasswd".to_string(),
-                        status,
-                    })
-                }
-            }
+
+    if password.is_empty() {
+        let status =
+            Command::new(path_passwd).arg("-d").arg(username).status()?;
+        if !status.success() {
+            return Err(Error::SubprocessFailed {
+                command: "passwd".to_string(),
+                status,
+            });
         }
+    } else {
+        // creating user with a non-empty password is not allowed.
+        return Err(Error::NonEmptyPassword);
     }
+
+    Ok(0)
 }
-impl From<&str> for Distributions {
-    fn from(s: &str) -> Self {
-        match s {
-            "debian" => Distributions::Debian,
-            "ubuntu" => Distributions::Ubuntu,
-            _ => panic!("Unknown distribution"),
-        }
+
+pub fn set_hostname(
```

Comment:
This is just an idea, and I've not thought too hard about it, but perhaps we should add some indirection here.

This is the `hostnamectl` implementation, and while I can't think of a current distro that doesn't use it, maybe some day that'll change and we should set the API up so we can accommodate that. One option would be to require selection of a single implementation at build time via feature flags or something, but another option would be to have the tool probe the known backends until one succeeds, or allow the caller to specify the backend (ultimately wired up to a CLI flag or environment variable, whatever the interface using this library wants).

If we go with the second approach, we could have an enumeration of implementations for hostname setting and have the API be something like
```
pub fn set_hostname(backend: backends::Hostname::HostnameCtl, hostname: &str) -> Result<i32, Error>
```
or, if we want to let the user specify multiple backends to try with ordered preference
```
pub fn set_hostname(backends: &[backends::Hostname], hostname: &str) -> Result<i32, Error>
```

Somewhat unrelated, we might want to map non-zero exit codes to an `Error` type and change the return to `Result<(), Error>` since as it is you have to check the result for an error, and then check the `Ok` value for an error.

Reply:
@jeremycline Thanks for the great suggestion.
But as I have dropped support for runtime configuration, the comment became outdated.